src/internal/observable/generate.ts

```diff
@@ -281,99 +281,99 @@ export function generate<T, S>(initialStateOrOptions: S | GenerateOptions<T, S>,
     resultSelector = resultSelectorOrObservable as ResultFunc<S, T>;
   }
 
-  return new Observable<T>(subscriber => {
-    let state = initialState;
-    if (scheduler) {
-      return scheduler.schedule<SchedulerState<T, S>>(dispatch, 0, {
+  if (scheduler) {
+    return new Observable(subscriber => {
+      const work = (state: SchedulerState<T, S>) => {
+        const { subscriber, condition } = state;
+        if (subscriber.closed) {
+          return;
+        }
+        if (state.needIterate) {
+          try {
+            state.state = state.iterate(state.state);
+          } catch (err) {
+            subscriber.error(err);
+            return;
+          }
+        } else {
+          state.needIterate = true;
+        }
+        if (condition) {
+          let conditionResult: boolean;
+          try {
+            conditionResult = condition(state.state);
+          } catch (err) {
+            subscriber.error(err);
+            return;
+          }
+          if (!conditionResult) {
+            subscriber.complete();
+            return;
+          }
+          if (subscriber.closed) {
+            return;
+          }
+        }
+        let value: T;
+        try {
+          value = state.resultSelector(state.state);
+        } catch (err) {
+          subscriber.error(err);
+          return;
+        }
+        if (subscriber.closed) {
+          return;
+        }
+        subscriber.next(value);
+        if (subscriber.closed) {
+          return;
+        }
+        subscriber.add(scheduler.schedule(work, 0, state));
```

Comment:
Could/should this use `reschedule`?

Reply:
I think I'd like to refactor these individually after this PR. It's already doing so much.